python/ray/air/_internal/torch_utils.py

```diff
@@ -292,3 +293,223 @@ def consume_prefix_in_state_dict_if_present_not_in_place(
+        batch,
+        dtypes=dtypes,
+        device=device,
+    )
+
+
+@torch.no_grad()
+def concat_tensors_to_device(
+    tensor_list: List[torch.Tensor],
+    device: str,
+) -> torch.Tensor:
+    """Stack list of tensors into a contiguous GPU tensor.
+
+    Args:
+        tensor_list: List of tensors to stack
+        device: The device to move tensors to
+
+    Returns:
+        A contiguous tensor on the target device
+    """
+    # Assumes tensors have the same shape/dtype
+    assert tensor_list, "Cannot stack empty list of tensors"
+    assert all(
+        isinstance(t, torch.Tensor) for t in tensor_list
+    ), "All items must be torch.Tensor"
+    assert all(
+        t.dtype == tensor_list[0].dtype for t in tensor_list
+    ), "All tensors must have the same dtype"
+    assert all(
+        t.shape[1:] == tensor_list[0].shape[1:] for t in tensor_list
+    ), "All tensors must have the same shape[1:]"
+
+    first = tensor_list[0]
+    dtype = first.dtype
+    shape_tail = first.shape[1:]
+    total_rows = sum(t.shape[0] for t in tensor_list)
+
+    # Allocate an empty Tensor on device
+    result = torch.empty((total_rows, *shape_tail), dtype=dtype, device=device)
+
+    row_start = 0
+    for t in tensor_list:
+        row_end = row_start + t.shape[0]
+        result[row_start:row_end].copy_(t, non_blocking=t.is_pinned())
+        row_start = row_end
+
+    assert isinstance(result, torch.Tensor), "Result must be a torch.Tensor"
+    return result
+
+
+@torch.no_grad()
+def move_tensors_to_device(
+    batch: Union[
+        torch.Tensor,
+        List[torch.Tensor],
+        Dict[str, torch.Tensor],
+        Dict[str, List[torch.Tensor]],
+    ],
+    device: Optional[str] = None,
+) -> Union[torch.Tensor, Dict[str, torch.Tensor],]:
+    """Move tensors to the specified device.
+
+    Args:
+        batch: A tensor or collection of tensors to move to device. Can be:
+            - A single tensor
+            - A list of tensors
+            - A dict mapping keys to tensors
+            - A dict mapping keys to lists of tensors
+        device: The device to move tensors to. If None, tensors are not moved.
+
+    Returns:
+        The input tensors moved to the specified device, maintaining the same structure.
+        Note that any lists of tensors will be concatenated into a single tensor.
+    """
+    if device is None:
+        return batch
+
+    if isinstance(batch, dict):
+        for k, v in batch.items():
+            if isinstance(v, list):
+                if all(isinstance(t, torch.Tensor) for t in v):
+                    batch[k] = concat_tensors_to_device(v, device=device)
+                else:
+                    raise TypeError(
+                        f"Expected list of torch.Tensor for key '{k}', got {[type(t) for t in v]}"
+                    )
+            elif isinstance(v, torch.Tensor):
+                batch[k] = v.to(device=device, non_blocking=v.is_pinned())
```

Comment:
I found `v.is_pinned()` was expensive in my perf profiling. 
We can just pass in a `non_blocking` flag